rbd-mirror: apply image state during snapshot replay

dillaman opened the pull request:
<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

dillaman commented:
GitHub fail -- I most certainly did not request a review from `ceph-volume`

trociny commented on src/librbd/mirror/snapshot/Types.h:
```diff
@@ -36,28 +36,24 @@ struct ImageStateHeader {
 WRITE_CLASS_ENCODER(ImageStateHeader);
 
 struct SnapState {
-  uint64_t id = CEPH_NOSNAP;
```
@dillaman The reason why I initially added the `id` field was that I was thinking to use it for remote-local snap_seq mapping. I.e. on the primary site it would be the same as the index, but on the secondary site, the rbd-mirror would set this field to the local snap id.

But as you decided to store the snap_seq mapping in the snapshot namespace, it is not relevant any more.

dillaman replied:
Gotcha -- yeah, we needed to store snap mappings for all snapshots between mirror snapshots so I needed `snap_seqs`.

trociny commented:
@dillaman There rbd_mirror test failures on teuthology [1] (and reproduced locally) like this one:
```
[ RUN      ] TestImageReplayer/1.UpdateFeatures
seed 22552
flushed
flushed
/home/mgolub/ceph/ceph/src/test/rbd_mirror/test_ImageReplayer.cc:1428: Failure
Expected equality of these values:
  (1ULL<<3) | (1ULL<<4)
    Which is: 24
  local_image_ctx->features & ( (1ULL<<3) | (1ULL<<4))
    Which is: 0
[  FAILED  ] TestImageReplayer/1.UpdateFeatures, where TypeParam = rbd::mirror::TestImageReplayerParams<(cls::rbd::MirrorImageMode)1> (21710 ms)
```
In debug logs I see:
```
2020-02-21T10:43:46.283+0200 7f2396ffd700 15 rbd::mirror::image_replayer::snapshot::ApplyImageStateRequest: 0x7f238c0630e0 ApplyImageStateRequest: image_state=[name=image1, features=317, snap_limit=18446744073709551615, snaps={5=[namespace=[trash original_name=.mirror.primary.ccf7e393-da05-4e22-af86-df767a9e4276.e5e13321-68dd-456a-a7e4-fbf407f1d61b, original_snapshot_namespace=mirror], name=c302e688-22d6-47b7-a324-8ec9ccacd35e, protection=0]}, metadata_count=0]
2020-02-21T10:43:46.283+0200 7f2396ffd700 15 rbd::mirror::image_replayer::snapshot::ApplyImageStateRequest: 0x7f238c0630e0 compute_remote_snap_id: local snapshot 4 maps to remote snapshot 5
2020-02-21T10:43:46.283+0200 7f2396ffd700 15 rbd::mirror::image_replayer::snapshot::ApplyImageStateRequest: 0x7f238c0630e0 compute_remote_snap_id: local snapshot 5 maps to remote snapshot 6
2020-02-21T10:43:46.283+0200 7f2396ffd700 15 rbd::mirror::image_replayer::snapshot::ApplyImageStateRequest: 0x7f238c0630e0 compute_remote_snap_id: local snapshot 6 maps to remote snapshot 8
2020-02-21T10:43:46.283+0200 7f2396ffd700 15 rbd::mirror::image_replayer::snapshot::ApplyImageStateRequest: 0x7f238c0630e0 compute_local_to_remote_snap_ids: local_to_remote_snap_ids={4=5,5=6,6=8}
2020-02-21T10:43:46.283+0200 7f2396ffd700 15 rbd::mirror::image_replayer::snapshot::ApplyImageStateRequest: 0x7f238c0630e0 update_features: image_features=37, state_features=317, feature_updates=280, enabled=1
2020-02-21T10:43:46.303+0200 7f23caffd700 -1 librbd::EnableFeaturesRequest: failed to update features: (22) Invalid argument
2020-02-21T10:43:46.303+0200 7f23caffd700 15 rbd::mirror::image_replayer::snapshot::ApplyImageStateRequest: 0x7f238c0630e0 handle_update_features: r=-22
```
It looks to me like a race between "create mirror primary snapshot" running by the test (flush) and "unlink peer" running by the Replayer for a replayed snapshot. The "unlink peer" triggers snapshot remove and "create mirror primary snapshot" may collect image state at the moment when the snapshot is moved in trash and the feature operation bit (1<<8) is set, which the rbd-mirror tries to replay. It looks like we need to filter out internal features?

And in `mirror:snapsho::SetImageStateRequest` may be we should filter out not only mirror snapshot but also trashed mirror snapshots?

[1] http://pulpito.ceph.com/trociny-2020-02-20_20:50:14-rbd-wip-mgolub-testing-distro-basic-smithi/

dillaman commented:
> And in `mirror:snapsho::SetImageStateRequest` may be we should filter out not only mirror snapshot but also trashed mirror snapshots?

I had actually already done that here [1] so I'll pull it over to this PR. I'll also include the internal feature bit filter. Thx.


[1] https://github.com/ceph/ceph/pull/33440/files#diff-ecba8030b635fc85aeb268dc4b7f8875R111

dillaman commented:
Turns out that fix was originally part of this branch but I dropped it somehow -- along w/ the mock unit tests for `ApplyImageStateRequest`. Updated.

dillaman commented:
FYI - I'm investigating a sporadic failure in `TestImageReplayer/1.SnapshotLimit` right now

dillaman commented:
OK, turns out that `SetImageStateRequest` was incorrectly being executed for non-primary snapshots and therefore marking the snapshots are prematurely complete. Fixed in the last commit.
